backend/fms_core/template_importer/row_handlers/project_link_samples/project_link_samples.py

```diff
@@ -28,18 +32,94 @@ def process_row_inner(self, sample, project, action):
+                        else:
+                            self.errors['study'] = f"Sample [{sample_obj.name}] is already queued in study [{project['study_letter']}] \
+                                of project [{project['name']}] at step [{project['step_order']}]"
+                    else:
+                        self.errors['study'] = f"Specified study [{project['study_letter']}] doesn't exist for project [{project['name']}]"
+                else:
+                    self.errors['add_to_study'] = f"A study needs to be specified to be able to add the sample to a study."
+
+
+            # Perform a remove project action
+            elif action['name'] == REMOVE_PROJECT_ACTION:
+                # If a study letter is provided
+                if project['study_letter']:
+                    self.errors['remove_project'] = f"A study [{project['study_letter']}] was specified. \
+                        If you intend to remove sample [{sample_obj.name}] from a single study use the [Remove from study] action."
+                # If the study is not specified, dequeue the sample from all the studies of the selected project
+                else:
+                    num_dequeued = 0
+                    for study in project_obj.studies.all():
+                         # Check if sample is queued 
+                        is_sample_queued, self.errors['sample_queued'], self.warnings['sample_queued'] = is_sample_queued_in_study(sample_obj, study)
+                        
+                        if is_sample_queued:
+                            # Don't specify order to be able remove all instances of sample in the study, regardless of step
+                            num_dequeued, self.errors['remove_from_study'], self.warnings['queue_to_study'] = dequeue_sample_from_all_steps_study_workflow(sample_obj, study_obj)
+                            num_dequeued += num_dequeued
+                    if num_dequeued > 0:
+                        self.warnings['remove_project'] = f"Sample [{sample_obj.name}] was removed from all the studies belonging to project [{project_obj.name}]"
+                
+                if not any(self.errors.values()):
+                    # Remove link object if no errors
+                    link_removed, self.errors['link'], self.warnings['link'] = remove_link(sample=sample_obj,
+                                                                                        project=project_obj)
+                
+
+            # Perform a remove study action
+            elif action['name'] == REMOVE_STUDY_ACTION:
+                # Dequeue sample from a specified study 
+                if project['study_letter']:
+                    study_obj, self.errors['study'], self.warnings['study'] = get_study(project_obj, project['study_letter'])
+
+                    if study_obj:
+                        # To avoid empty step orders
+                        step_order = project['step_order'] if project['step_order'] else 1
```

Comment:
ick ... no ... i would prefer we make it impossible for users to remove sample from completed study than assuming that an empty field means 1 ;). I would raise an error for remove_study if the order is empty. Removing a sample that completed a study will be possible from the direct service call for us and we can do curations otherwise (i don't see any use case to do that for the lab).